### ISSUE TITLE
vote hash mismatch on mainter against mb

### ISSUE THREAD

**Issue by ryoqun** [solana-labs/solana#17083]
#### Problem

After several ~ 24 hours, master's `solana-validator` forks off from mb. v1.6 branch isn't affected:

BAD

```
$ grep  -A1 bank.*rozen.*7705208[10].*cap solana-validator-ALRTjMiAkxNNzTeRho1iW3hPMr9bRawicW2hdLDYe1sv.log.2021y05m06d02h30m49s419451345ns
[2021-05-06T04:30:48.681506333Z INFO  solana_runtime::bank] bank frozen: 77052080 hash: Ez7JguTesdhMeJXbCy2Pqe3MEAyTcXP9jJAhwr1dsU9c accounts_delta: Db56a62B3RTfYAPUHrYsTQhMsXYoAnUo5GxJD4QoyFbF signature_count: 1222 last_blockhash: 5LHwb6pw3N7CrTNtRoaqtJoD7jWK1EL8H1GTNBgPNtLu capitalization: 494729633919310853
[2021-05-06T04:30:48.681543930Z INFO  solana_runtime::bank] accounts hash slot: 77052080 stats: BankHashStats { num_updated_accounts: 3510, num_removed_accounts: 2, num_lamports_stored: 12250358615251674, total_data_len: 411977746, num_executable_accounts: 2 }
--
[2021-05-06T04:30:49.109197394Z INFO  solana_runtime::bank] bank frozen: 77052081 hash: 55qC5WaiTeCdz7iRep42XLWqQRUXjNh4epTuNHYtZLed accounts_delta: AoQfYPepNACVDQhrTW4FHrpKQA2zssJzFUVfVy6ZHbFW signature_count: 494 last_blockhash: C5Ff3EhB9zeUbdfiXtNpnhsoBKwhwCTwWfdn8igWXHmA capitalization: 494729633918075853
[2021-05-06T04:30:49.109220784Z INFO  solana_runtime::bank] accounts hash slot: 77052081 stats: BankHashStats { num_updated_accounts: 1759, num_removed_accounts: 5, num_lamports_stored: 7984299796894355, total_data_len: 276549138, num_executable_accounts: 3 }

```

GOOD

```
$ grep -A1 bank.*rozen.*7705208[10].*cap solana-validator-21XVxUs4s1zSgkGNWgQDvTkNwKwgAWYHwYaiPyrr3HLA.log
[2021-05-06T04:30:48.605019353Z INFO  solana_runtime::bank] bank frozen: 77052080 hash: Ez7JguTesdhMeJXbCy2Pqe3MEAyTcXP9jJAhwr1dsU9c accounts_delta: Db56a62B3RTfYAPUHrYsTQhMsXYoAnUo5GxJD4QoyFbF signature_count: 1222 last_blockhash: 5LHwb6pw3N7CrTNtRoaqtJoD7jWK1EL8H1GTNBgPNtLu capitalization: 494729633919310853
[2021-05-06T04:30:48.605051351Z INFO  solana_runtime::bank] accounts hash slot: 77052080 stats: BankHashStats { num_updated_accounts: 3510, num_removed_accounts: 2, num_lamports_stored: 12250358615251674, total_data_len: 411977746, num_executable_accounts: 2 }
--
[2021-05-06T04:30:49.115992893Z INFO  solana_runtime::bank] bank frozen: 77052081 hash: 37UhwEWabghsD14tjSjY65SMhQg4hRABu7hYCxpXfP5M accounts_delta: 4CovMT53n9x2TnfswtvGkoYFrRASnhqbnKmF9foktWUY signature_count: 494 last_blockhash: C5Ff3EhB9zeUbdfiXtNpnhsoBKwhwCTwWfdn8igWXHmA capitalization: 494729633918075853
[2021-05-06T04:30:49.116017020Z INFO  solana_runtime::bank] accounts hash slot: 77052081 stats: BankHashStats { num_updated_accounts: 1759, num_removed_accounts: 5, num_lamports_stored: 7984299796894355, total_data_len: 276549138, num_executable_accounts: 3 }
```

Surprisingly, logs doesn't tell.. it seems there's very subtle difference.

**Comment by ryoqun** [comment]
I've just chased down this into these bad snapshots:

```
$ ls -l /tmp/*77052081*
-rw-rw-r-- 1 sol sol 6597048055 May  6 06:44 /tmp/snapshot-77052081-Hkvfo8eL4AdqYZj9PuofVUrARUdLZ6GFiGEvSvFiP4iy.tar.zst
-rw-rw-r-- 1 sol sol 6596634977 May  6 07:04 /tmp/snapshot-77052081-TG182yxZBv9PCZbN2Q2CwWHPAnRgho68eiot5xfaWuL.tar.zst

```

**Comment by ryoqun** [comment]
@jackcmay I haven't looked in detail. but it looks like there is some new change in upgradable loader? Is there something to ring?

```
--- good.yml    2021-05-06 08:23:20.853776363 -0700
+++ bad.yml     2021-05-06 08:23:30.130609670 -0700
@@ -924,21 +924,21 @@
   executable: false
   rentEpoch: 178
 5KKsLVU6TcbVDK4BS6K1DGDxnh4Q9xjYJ8XaDCG5t8ht:
   lamports: 1825496640
   data: c2VydW0RAAAAAAAAALYBAAAAAAAAAAAAAAAAAABMCicAAAAAAAkCAwAAAAAA28bkDgAAAAAAOMlNAQAAALskAQAAAAAAJBwWAAAAAAA/rgAAAAAAADVy2O6wzLdRE5EHksK09JIKFvhUuwJCtlKV9TqRHDUEJ/phMt3MqtYFAAAAAAAAAAA4yU0BAAAA1wTsDgAAAAC3YggAAAAAAMrj6f//////H68AAAAAAADvOWpbyu1uPaoRLcBr3Hlb2r4jxRCnPcWUIF7rHlOr/wAAAAAAAAAAFgAAAAAAAABJwQoAAAAAAAAAAAAAAAAAAAAAAAAAAADK4+n//////x+vAAAAAAAA7>
   owner: 9xQeWvG816bUx9EPjHmaT23yvVM2ZWbrrpZb9PusVFin
   executable: false
   rentEpoch: 178
 5KW2RMYEAwr38GXNMa3PPe22sa2rbFjb9DFjCgegiT4H:
   lamports: 10613255280
-  data: AwAAALG4lwQAAAAAARVn1V6H4RbCMI4IW2PMx3VQ8YyZ9cpb1+3QEvyoWA6sf0VMRgIBAQAAAAAAAAAAAAMA9wABAAAAiOUAAAAAAABAAAAAAAAAABAhDQAAAAAAAAAAAEAAOAADAEAADAALAAEAAAAFAAAA6AAAAAAAAADoAAAAAAAAAOgAAAAAAAAA4LkMAAAAAADguQwAAAAAAAAQAAAAAAAAAQAAAAYAAADQugwAAAAAANC6DAAAAAAA0LoMAAAAAAAQZQAAAAAAABBlAAAAAAAAABAAAAAAAAACAAAABgAAAOAfDQAAAAAA4B8NAAAAAADgHw0AAAAAANAAAAAAAAAA0>
+  data: AwAAADgGlwQAAAAAARVn1V6H4RbCMI4IW2PMx3VQ8YyZ9cpb1+3QEvyoWA6sf0VMRgIBAQAAAAAAAAAAAAMA9wABAAAAiOUAAAAAAABAAAAAAAAAAKgLDQAAAAAAAAAAAEAAOAADAEAADAALAAEAAAAFAAAA6AAAAAAAAADoAAAAAAAAAOgAAAAAAAAAuKQMAAAAAAC4pAwAAAAAAAAQAAAAAAAAAQAAAAYAAACgpQwAAAAAAKClDAAAAAAAoKUMAAAAAADYZAAAAAAAANhkAAAAAAAAABAAAAAAAAACAAAABgAAAHgKDQAAAAAAeAoNAAAAAAB4Cg0AAAAAANAAAAAAAAAA0>
   owner: BPFLoaderUpgradeab1e11111111111111111111111
   executable: false
   rentEpoch: 178
 5KgKdCWVyWi9YJ6GipzozhWxAvnbQPpUtaxuMXXEn3Zs:
   lamports: 2039280
   data: Ii9B9QktgU3mZCAt2+S0ROmkSIWPpH3vtyadWtu/wgRPREsEBR92YmUjRbotO5hrTnWGfGp3cCkXj8nG57nMTqx3wQEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
   owner: TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA
   executable: false
   rentEpoch: 178
 5LmHe3x8VwGzWZ6rooARZJNMo6AaN1P73478AuhBUjUr:

```

**Comment by ryoqun** [comment]
Excerpt of diff of decoded account data.

Hmm, my eyes aren't not trained for ELF.. But the diff is large and it looks like this is reverted state or something?

```
--- good-account-hex    2021-05-06 08:38:19.382474461 -0700
+++ bad-account-hex     2021-05-06 08:38:51.209333342 -0700
@@ -1,32 +1,32 @@
-00000000: 0300 0000 b1b8 9704 0000 0000 0115 67d5  ..............g.
+00000000: 0300 0000 3806 9704 0000 0000 0115 67d5  ....8.........g.
 00000010: 5e87 e116 c230 8e08 5b63 ccc7 7550 f18c  ^....0..[c..uP..
 00000020: 99f5 ca5b d7ed d012 fca8 580e ac7f 454c  ...[......X...EL
 00000030: 4602 0101 0000 0000 0000 0000 0003 00f7  F...............
 00000040: 0001 0000 0088 e500 0000 0000 0040 0000  .............@..
-00000050: 0000 0000 0010 210d 0000 0000 0000 0000  ......!.........
+00000050: 0000 0000 00a8 0b0d 0000 0000 0000 0000  ................
 00000060: 0040 0038 0003 0040 000c 000b 0001 0000  .@.8...@........
 00000070: 0005 0000 00e8 0000 0000 0000 00e8 0000  ................
-00000080: 0000 0000 00e8 0000 0000 0000 00e0 b90c  ................
-00000090: 0000 0000 00e0 b90c 0000 0000 0000 1000  ................
-000000a0: 0000 0000 0001 0000 0006 0000 00d0 ba0c  ................
-000000b0: 0000 0000 00d0 ba0c 0000 0000 00d0 ba0c  ................
-000000c0: 0000 0000 0010 6500 0000 0000 0010 6500  ......e.......e.
+00000080: 0000 0000 00e8 0000 0000 0000 00b8 a40c  ................
+00000090: 0000 0000 00b8 a40c 0000 0000 0000 1000  ................
+000000a0: 0000 0000 0001 0000 0006 0000 00a0 a50c  ................
+000000b0: 0000 0000 00a0 a50c 0000 0000 00a0 a50c  ................
+000000c0: 0000 0000 00d8 6400 0000 0000 00d8 6400  ......d.......d.
 000000d0: 0000 0000 0000 1000 0000 0000 0002 0000  ................
-000000e0: 0006 0000 00e0 1f0d 0000 0000 00e0 1f0d  ................
-000000f0: 0000 0000 00e0 1f0d 0000 0000 00d0 0000  ................
+000000e0: 0006 0000 0078 0a0d 0000 0000 0078 0a0d  .....x.......x..
+000000f0: 0000 0000 0078 0a0d 0000 0000 00d0 0000  .....x..........
 00000100: 0000 0000 00d0 0000 0000 0000 0008 0000  ................
 00000110: 0000 0000 0079 1100 0000 0000 0061 1300  .....y.......a..
 00000120: 0000 0000 0055 0303 0001 0000 0007 0100  .....U..........
-00000130: 0004 0000 0085 1000 0074 8300 0005 0002  .........t......
+00000130: 0004 0000 0085 1000 002f 8300 0005 0002  ........./......
 00000140: 0000 0000 0007 0100 0008 0000 0085 1000  ................
-00000150: 00e7 5a01 0095 0000 0000 0000 0095 0000  ..Z.............
+00000150: 0099 5801 0095 0000 0000 0000 0095 0000  ..X.............
 00000160: 0000 0000 00bf 1600 0000 0000 0061 6100  .............aa.
 00000170: 0000 0000 0015 0101 000e 0000 0095 0000  ................
 00000180: 0000 0000 0007 0600 0008 0000 00bf 6100  ..............a.
-00000190: 0000 0000 0085 1000 0090 1d01 00bf 6100  ..............a.
-000001a0: 0000 0000 0085 1000 0069 dc00 0005 00f9  .........i......
+00000190: 0000 0000 0085 1000 005d 1b01 00bf 6100  .........]....a.
+000001a0: 0000 0000 0085 1000 000e da00 0005 00f9  ................
 000001b0: ff00 0000 007b 2100 0000 0000 0067 0300  .....{!......g..
 000001c0: 0005 0000 000f 3200 0000 0000 007b 2108  ......2......{!.
 000001d0: 0000 0000 0095 0000 0000 0000 0079 2320  .............y# 
 000001e0: 0000 0000 007b 3120 0000 0000 0079 2318  .....{1 .....y#.
 000001f0: 0000 0000 007b 3118 0000 0000 0079 2310  .....{1......y#.
@@ -37,1093 +37,1093 @@
```

**Comment by ryoqun** [comment]
https://explorer.solana.com/block/77052081

**Comment by ryoqun** [comment]
... and supposedly this is the problematic transaction because this is the only `BPFLoaderUpgradeab1e11111111111111111111111` transaction : https://explorer.solana.com/tx/23Nqqy331GoaWS2UMnncv4dyPjZAcepM7PqpJdngpyPnoSCyLHDq1Wm3gmzmRrJsYnbW2ewXJUco4pjLumFPEMA4

**Comment by jackcmay** [comment]
@ryoqun Thanks for the sluthing!  Are you seeing both mainnet and master transactions succeed, but just with different account data?  Or are you seeing it fail in one place and pass in the other?

@Lichtso I'm down for the count today and looking over the diffs between v1.6 and master the big change to the upgradeable loader is KeyedAccount rework.  Would you be able to take a look?

**Comment by ryoqun** [comment]
> @ryoqun Thanks for the sluthing! Are you seeing both mainnet and master transactions succeed, but just with different account data? Or are you seeing it fail in one place and pass in the other?

@jackcmay 

> just with different account data?

This is the case. Both v1.6 node and master node processed the tx as success but resulted in different account data (maybe this account contains ELF?) So, I think we can reproduce this with local or testnet by deploying a program. Also `solana-ledger-tool` block processing (replaying) also exposes the bug. So, that should be easy to repro locally. :)

**Comment by ryoqun** [comment]
@mvines just headsup, I think this is a v1.7 blocker for testnet. This bug seems to introduce some not-deterministic behavior for upgradeable program deploy.

**Comment by jackcmay** [comment]
@ryoqun Are you mention that v1.6 isn't affected, does this mean that you are seeing the issue between master and v1.5?

**Comment by jackcmay** [comment]
This is a strange one, this is an upgrade via CPI and both succeed but with different ELF data.   I don't see anything obvious between master and either v1.6 or v1.5 that would cause this.  Because this is an upgrade via CPI it is possible that the program modified the ELF data before calling upgrade and maybe it has to do with getting the caller's keyed accounts from invoke_context rather than from the last loader invoke?  Since we have the snapshots, @ryoqun would you recommend the best way to debug this is to reply that snapshot on both branches with more debug info added?

**Comment by ryoqun** [comment]
Err, I meant v1.6 doesn't have this regression bug. The master branch seems to have this bug. I haven't tested v1.5 at all to begin with because it's effectively EOL-ed.
I had some difficuly to reproduce this non-deterministic behavior with run.sh by simply `solana program deploy`...
@jackcmay  So, I created minimal version of mainnet-beta ledger. (Still 6G though, so takes a bit to download i guess....) 

> @ryoqun would you recommend the best way to debug this is to reply that snapshot on both branches with more debug info added?

Yeah, sadly, this will be the best way...

https://drive.google.com/drive/folders/1brIVbrUQnv34pg7_UeT1DPHB27-A1YmE?usp=sharing

```
solana-ledger-tool --version
solana-ledger-tool 1.6.6 (src:fe775a97; feat:3714435735)
solana-ledger-tool --ledger bad-deploy-ledger-copy/ create-snapshot 77052081 created-snapshots/ &> bad-deploy-ledger-copy-77052081-good.log
```

note that b295a0d7 is in my pr (#16697 )

```
solana-ledger-tool --version
solana-ledger-tool 1.7.0 (src:b295a0d7; feat:3458834192)
solana-ledger-tool --ledger bad-deploy-ledger-copy/ create-snapshot 77052081 created-snapshots/ &> bad-deploy-ledger-copy-77052081-bad.log
```

Then this should create these differing snapshots:

```
$ ls -l created-snapshots/
total 12884640
-rw-rw-r-- 1 sol sol 6600323462 May  9 09:35 snapshot-77052081-Hkvfo8eL4AdqYZj9PuofVUrARUdLZ6GFiGEvSvFiP4iy.tar.zst
-rw-rw-r-- 1 sol sol 6593524134 May  9 09:47 snapshot-77052081-TG182yxZBv9PCZbN2Q2CwWHPAnRgho68eiot5xfaWuL.tar.zst
```

**Comment by ryoqun** [comment]
> https://drive.google.com/drive/folders/1brIVbrUQnv34pg7_UeT1DPHB27-A1YmE?usp=sharing

@jackcmay Sorry, I put broken rocksdb in the original whole archive. So, I uploaded another one: `bad-deploy-ledger-copy-fixed-rocks-only.tar.bz2`...

**Comment by Lichtso** [comment]
### Transaction History of the Program's Upgrade Authority

Transaction | Slot | Date | Success
-- | -- | -- | --
[23Nqqy331GoaWS2UMnncv4dyPjZAcepM7PqpJdngpyPnoSCyLHDq1Wm3gmzmRrJsYnbW2ewXJUco4pjLumFPEMA4](https://explorer.solana.com/tx/23Nqqy331GoaWS2UMnncv4dyPjZAcepM7PqpJdngpyPnoSCyLHDq1Wm3gmzmRrJsYnbW2ewXJUco4pjLumFPEMA4) | [77052081](https://explorer.solana.com/block/77052081) | 4 days ago | ✔️
[55TjJ8yhw49X8pRjyhY9MBtd2XQLGU3QBKnnpsxEjtVGiQ8QPzeivBdT79XazJZYDbzagPbX5hkCB4se9Aic4LuF](https://explorer.solana.com/tx/55TjJ8yhw49X8pRjyhY9MBtd2XQLGU3QBKnnpsxEjtVGiQ8QPzeivBdT79XazJZYDbzagPbX5hkCB4se9Aic4LuF) | [77051050](https://explorer.solana.com/block/77051050) | 4 days ago | ✔️
[4gsZCMgLndwzEZwdu7CZdVEzxQgfnaxGubZ3RSL7gCJNikEsTeUJZ3hJrro19hjec9Jcv5BMnW7cAWkw4cmQ3pT6](https://explorer.solana.com/tx/4gsZCMgLndwzEZwdu7CZdVEzxQgfnaxGubZ3RSL7gCJNikEsTeUJZ3hJrro19hjec9Jcv5BMnW7cAWkw4cmQ3pT6) | [77006392](https://explorer.solana.com/block/77006392) | 5 days ago | ✔️



### Decoded Header of the Account Diff:
`=0300 0000`: UpgradeableLoaderState=[ProgramData](https://github.com/solana-labs/solana/blob/22c02b917e866d00a93d20dd651af91cde5f117e/sdk/program/src/bpf_loader_upgradeable.rs#L39)
`-b1b8 9704 0000 0000`: slot=[77052081](https://explorer.solana.com/block/77052081)
`+3806 9704 0000 0000`: slot=[77006392](https://explorer.solana.com/block/77006392)
`=0000 0000 01`: [Some()](https://doc.rust-lang.org/std/option/enum.Option.html)
`=1567 d55e 87e1 16c2 308e 085b 63cc c775 50f1 8c99 f5ca 5bd7 edd0 12fc a858 0eac`: Pubkey=[2SZPmP91s1beN2MHwn7zhkfbBEGCgdVFvPB2rA9T9o3Z](https://explorer.solana.com/address/2SZPmP91s1beN2MHwn7zhkfbBEGCgdVFvPB2rA9T9o3Z)
`=7f45 4c46 0201 0100 0000 0000 0000 0000 0300 f700 0100 0000 88e5 0000 0000 0000 4000 0000 0000 0000 ...`: ELF

**Comment by jackcmay** [comment]
Hmm, interesting.  So the instruction was successful on both branches from on master the account got rolled back to two upgrades ago (back 45689 slots).  This is probably not specific to the loader but instead, something is funny in the account database?

**Comment by ryoqun** [comment]
@Lichtso hi, let's have a fun volleyball play. :) I'm pretty sure this vote hash mismatch is exactly caused by the commit 9dfcb921cf.. (I bisected..) Could you take a look at this as the next turn? Also, please let me know if you're having trouble reproduce this with steps https://github.com/solana-labs/solana/issues/17083#issuecomment-835847972 and  https://github.com/solana-labs/solana/issues/17083#issuecomment-836232869

```
$ git log --oneline --graph --decorate=full 610fa60d bc0ff373 | head
* bc0ff37387 (HEAD -> refs/heads/for-small-patches, tag: refs/tags/bad-77052081) avoid assert!?
* 9dfcb921cf Refactoring: Move KeyedAccounts to InvokeContext (#15410)
| * 610fa60d3a (tag: refs/tags/good-77052081) save # avoid assert
| * dde9f0cd46 save                                                   # avoid assert
|/  
* 015bc034a5 improve failing assert (#16581)
* 1e638c1371 chore:(deps): bump @solana/spl-token-registry in /explorer (#16631)
* 376d0bf063 chore:(deps): bump @testing-library/user-event in /explorer (#16630)
* b06e93fe5b Increase test timeout
* a911ae00ba clippy

$ /tmp/solana-ledger-tool-before-keyed-account-refactor --version
solana-ledger-tool 1.7.0 (src:610fa60d; feat:3714435735)
... produced
-rw-rw-r-- 1 sol sol 6594024482 May 13 07:54 snapshot-77052081-Hkvfo8eL4AdqYZj9PuofVUrARUdLZ6GFiGEvSvFiP4iy.tar.zst

$ /tmp/solana-ledger-tool-after-keyed-account-refactor --version
solana-ledger-tool 1.7.0 (src:bc0ff373; feat:3714435735)
... produced

-rw-rw-r-- 1 sol sol 6597886429 May 13 08:29 snapshot-77052081-TG182yxZBv9PCZbN2Q2CwWHPAnRgho68eiot5xfaWuL.tar.zst

```

**Comment by ryoqun** [comment]
and this patch:

```
$ git log -n1 -p
commit bc0ff37387bd814d454de919a7118da182cd5c12 (HEAD -> for-small-patches, tag: bad-77052081)
Author: Ryo Onodera <ryoqun@gmail.com>
Date:   Thu May 13 23:10:03 2021 +0900

    avoid assert!?

diff --git a/runtime/src/accounts_index.rs b/runtime/src/accounts_index.rs
index d78fa87b81..ab2fbf8f2f 100644
--- a/runtime/src/accounts_index.rs
+++ b/runtime/src/accounts_index.rs
@@ -309,7 +309,7 @@ impl Default for RootsTracker {
         // we expect to keep a rolling set of 400k slots around at a time
         // 2M gives us plenty of extra room to handle a width 5x what we should need.
         // cost is 2M bits of memory
-        RootsTracker::new(2097152)
+        RootsTracker::new(2097152 * 2)
     }
 }

```

**Comment by jeffwashington** [comment]
@ryoqun [here](https://github.com/solana-labs/solana/issues/17135#issuecomment-839402312) and above.
I'm working with @Lichtso right now on this.

**Comment by Lichtso** [comment]
@ryoqun 
> I'm pretty sure this vote hash mismatch is exactly caused by the commit [9dfcb92](https://github.com/solana-labs/solana/commit/9dfcb921cfea2449a714af627eeb9b367c2f39c2).. (I bisected..) Could you take a look at this as the next turn?

We have narrowed it down even further inside the PR and found the following "fix":
https://github.com/Lichtso/solana/commits/fix/invoke_context_push_account_deps

But, I am still not sure why this changes it ... or if the two issues (#17083 and #17135) are actually caused by the same thing.

**Comment by ryoqun** [comment]
@Lichtso thanks for the info! it seems your fix actually fixes this vote hash mismatch as well... Indeed, it's odd to me as for the reason this fixes the bugs. ;)

```
/tmp/solana-ledger-tool-after-keyed-account-refactor-with-fix --version
solana-ledger-tool 1.7.0 (src:7726ff3c; feat:3714435735)
... created good snapshot:
-rw-rw-r-- 1 sol sol 6593723297 May 13 22:12 snapshot-77052081-Hkvfo8eL4AdqYZj9PuofVUrARUdLZ6GFiGEvSvFiP4iy.tar.zst


```


```
* 7726ff3c2d (HEAD -> refs/heads/for-small-patches, tag: refs/tags/fixed-77052081) Apply Lichtso's patch
* bc0ff37387 (tag: refs/tags/bad-77052081) avoid assert!?
* 9dfcb921cf Refactoring: Move KeyedAccounts to InvokeContext (#15410)
| * 610fa60d3a (tag: refs/tags/good-77052081) save
| * dde9f0cd46 save
|/  
* 015bc034a5 improve failing assert (#16581)
* 1e638c1371 chore:(deps): bump @solana/spl-token-registry in /explorer (#16631)
* 376d0bf063 chore:(deps): bump @testing-library/user-event in /explorer (#16630)
* b06e93fe5b Increase test timeout
* a911ae00ba clippy
```


```patch
commit 7726ff3c2dc73d52d6b9f90003aca19d188a5003
Author: Ryo Onodera <ryoqun@gmail.com>
Date:   Fri May 14 13:57:38 2021 +0900

    Apply Lichtso's patch

diff --git a/programs/bpf_loader/src/lib.rs b/programs/bpf_loader/src/lib.rs
index fab9defc98..9936553d60 100644
--- a/programs/bpf_loader/src/lib.rs
+++ b/programs/bpf_loader/src/lib.rs
@@ -309,8 +309,6 @@ fn process_loader_upgradeable_instruction(
             let buffer = keyed_account_at_index(keyed_accounts, 3)?;
             let rent = from_keyed_account::<Rent>(keyed_account_at_index(keyed_accounts, 4)?)?;
             let clock = from_keyed_account::<Clock>(keyed_account_at_index(keyed_accounts, 5)?)?;
-            // TODO [KeyedAccounts to InvokeContext refactoring]
-            // let _system = keyed_account_at_index(keyed_accounts, 6)?;
             let authority = keyed_account_at_index(keyed_accounts, 7)?;
             let upgrade_authority_address = Some(*authority.unsigned_key());
             let upgrade_authority_signer = authority.signer_key().is_none();
diff --git a/runtime/src/message_processor.rs b/runtime/src/message_processor.rs
index 8f774cfa78..24466a949e 100644
--- a/runtime/src/message_processor.rs
+++ b/runtime/src/message_processor.rs
@@ -353,12 +353,17 @@ impl<'a> InvokeContext for ThisInvokeContext<'a> {
                     .chain(self.message.account_keys.iter())
                     .position(|key| key == *search_key)
                     .map(|mut index| {
+                        // TODO
+                        // Currently we are constructing new accounts on the stack
+                        // before calling MessageProcessor::process_cross_program_instruction
+                        // Ideally we would recycle the existing accounts here.
                         if index < self.account_deps.len() {
                             (
                                 *is_signer,
                                 *is_writable,
                                 &self.account_deps[index].0,
-                                &self.account_deps[index].1 as &RefCell<AccountSharedData>,
+                                // &self.account_deps[index].1 as &RefCell<AccountSharedData>,
+                                transmute_lifetime(*account),
                             )
                         } else {
                             index = index.saturating_sub(self.account_deps.len());
@@ -366,10 +371,6 @@ impl<'a> InvokeContext for ThisInvokeContext<'a> {
                                 *is_signer,
                                 *is_writable,
                                 &self.message.account_keys[index],
-                                // TODO
-                                // Currently we are constructing new accounts on the stack
-                                // before calling MessageProcessor::process_cross_program_instruction
-                                // Ideally we would recycle the existing accounts here like this:
                                 // &self.accounts[index] as &RefCell<AccountSharedData>,
                                 transmute_lifetime(*account),
                             )

```

**Comment by Lichtso** [comment]
@ryoqun Thanks for verifying that.

We still need to figure out what situation / scenario specifically causes both of these (#17083 and #17135).
Because the tests don't cover it yet, but they need to.
One thing they both seem to have in common is the `UpgradeableLoader` writing to an account from `account_deps`.

**Comment by ryoqun** [comment]
@Lichtso Okay. the final #17350 is validated. Really declaring the complete fix

```
$ /tmp/solana-ledger-tool-after-keyed-account-refactor-with-fix-final --version
solana-ledger-tool 1.7.0 (src:d5c1848a; feat:3714435735)
$ ls -l created-snapshots2/
...
-rw-rw-r-- 1 sol sol 6593510503 May 24 01:50 snapshot-77052081-Hkvfo8eL4AdqYZj9PuofVUrARUdLZ6GFiGEvSvFiP4iy.tar.zst


$ git log --oneline --graph --decorate=full HEAD | head -n 20
* d5c1848ab1 (HEAD -> refs/heads/for-small-patches) Fix InvokeContext::push() account_deps (#17350)
* bc0ff37387 (tag: refs/tags/bad-77052081) avoid assert!?
* 9dfcb921cf Refactoring: Move KeyedAccounts to InvokeContext (#15410)
* 015bc034a5 improve failing assert (#16581)
* 1e638c1371 chore:(deps): bump @solana/spl-token-registry in /explorer (#16631)
* 376d0bf063 chore:(deps): bump @testing-library/user-event in /explorer (#16630)
* b06e93fe5b Increase test timeout
* a911ae00ba clippy
* 3b79b21e9d Upgrade to Rust 1.51.0
* cfe7a4340b adds a shrink policy to the recycler without an allocation limit
* e405747409 Revert "Add limit and shrink policy for recycler (#15320)"

```
